### ISSUE TITLE
Extend/reuse Asciidoctor.js CLI

### ISSUE THREAD

**Issue by Mogztter** [Mogztter/asciidoctor-pdf.js#19]
Currently we are using a fork of [Asciidoctor.js CLI](https://github.com/asciidoctor/asciidoctor-cli.js). More specifically the code is copied and pasted 😳 😞 

We should try to reuse and extend the Asciidoctor.js CLI as much as possible to avoid duplication and reduce maintenance.

**Comment by oncletom** [comment]
Now it is possible to do `asciidoctor --require asciidoctor-pdf ...`, do you think it is possible to "extend" the CLI with a subprogram? (commander and/or yargs enable that).

In a way, the Asciidoctor CLI interprets the arguments and provide them to `asciidoctor-pdf`.

One liner with `npx` would look like:

```bash
npx -p asciidoctor -p asciidoctor-pdf asciidoctor pdf -o sample.pdf --template-require template.js --stylesheet sample.css sample.adoc
```

What do you think?

**Comment by Mogztter** [comment]
Our minds are in sync! I was thinking about it yesterday and I think it's a great idea 🤯

Initially, I was thinking about using require with different "types of modules". I really want the CLI to be as consistent as possible so maybe we should leverage `--require`.

Currently we can require extensions and converters but we could introduce a "template module".
In my opinion `--template-dir` is not flexible enough in a Node.js context. It could be useful to provide a template module that could contain JavaScript files, dependencies, template files (Pug, Mustache...) organized in any way that make sense for the user.

But at the same time I really like the subprogram approach because it's more explicit.

I also like the `--stylesheet` option. I think one of the biggest issue right now is how to resolve external resources. Asciidoctor also has this issue and, if possible, I want to avoid the whole `base_dir`, `docdir`, `data-uri` shenanigans. While it's extremely powerful, I think it's also very confusing for newcomers. In this case, it will be even more confusing because we need to convert the document to an HTML file in an intermediate step. So it's really an internal thing and the user doesn't really need to know where the temporary HTML file is written.
Anyway we should probably eagerly resolve relative paths or enforce that paths are absolute and then embed everything in the HTML output so the file is portable. 
Extending this idea, it could also be useful in an offline situation if we cache remote assets or in a live-reload use case where you don't want your browser to constantly fetch an remote ressources.

In my opinion, this project should probably be splitted.
From a user point of view, it's just a "converter" and to be consistent we should be able to write:

```
asciidoctor --require @asciidoctor/pdf-converter -b pdf doc.adoc
```
```
asciidoctor --require @asciidoctor/docbook-converter -b docbook doc.adoc
asciidoctor --require @asciidoctor/reveal.js-converter -b reveal.js doc.adoc
asciidoctor --require @asciidoctor/manpage-converter -b manpage doc.adoc
asciidoctor --require @asciidoctor/opendocument-converter -b opendocument doc.adoc
```

At some point we might promote some converters to be able to write:

```
asciidoctor -b docbook doc.adoc
```
(In this case the require is implicit because the Asciidoctor CLI known that it needs to require `@asciidoctor/pdf-converter` when the requested backend is `pdf`)

Or if we want to use a subprogram:

```
asciidoctor docbook doc.adoc
```

I think @mojavelinux's idea was to create a dedicated CLI for each converters:

* `asciidoctor-bespoke`
* `asciidoctor-pdf`

To summarize, I think one part is the converter, another part is the CLI and a third part could be the default templates that we provide out-of-the-box. So if the user does not specify anything we should convert the document using a default template.

**Comment by Mogztter** [comment]
For reference, the Asciidoctor Bespoke binary uses the Asciidoctor CLI: https://github.com/asciidoctor/asciidoctor-bespoke/blob/master/bin/asciidoctor-bespoke

If we follow the same approach, the command would look like:

```
npx asciidoctor-pdf -o sample.pdf --template-require template.js --stylesheet sample.css sample.adoc
```

**Comment by oncletom** [comment]
I understand you need to use `-p` for each package you require to run your command. As of [call with multiple packages](https://www.npmjs.com/package/npx#execute-a-full-shell-command-using-one-npx-call-w-multiple-packages).

This works for me:

```bash
$ npx -p @asciidoctor/docbook-converter -p asciidoctor asciidoctor --require @asciidoctor/docbook-converter -b docbook5 sample.adoc
```

**Comment by Mogztter** [comment]
> This works for me:

That's great news, I will try with the latest version of npx.
I've created an issue in the Asciidoctor CLI repository to create a proper API: https://github.com/asciidoctor/asciidoctor-cli.js/issues/64

**Comment by oncletom** [comment]
This is the version installed on my machine.

```bash
$ npx --version
6.9.0
```
